File: test/go/grammar.go
[keyword package] [variable hello] [comment /**/]

[keyword import] [def m] [string "lib/math"]

[keyword func] [def generate]([def&local ch] [keyword chan][operator <-] [type int]) {
	[keyword for] [def&local i] [operator :=] [number 2]; ; [variable-2 i][operator ++] {
		[variable-2 ch] [operator <-] [variable-2 i]  [comment // Send 'i' to channel 'ch'.]
	}
}

[keyword const] (
	[def x], [def y] [type int]
	[def p] [[]][type Something]; [def u] [operator =] [string '\n']
)

[keyword type] [def j] [keyword struct] {
	[def&property x], [def&property y] [type int]
	[def&property u] [type float32]
	[def&property _] [type float32]  [comment // padding]
	[def&property A] *[[]][type int]
	[def&property F] [keyword func]()
}

[keyword if] [variable x] [operator +] [variable y] {
	[variable x][operator ++]
	[variable y] [operator -=] [variable x]; [variable&callee foo]([number 0xaffe])
} [keyword else] [keyword if] [atom true] {
	[keyword return] [string `plain]
[string stringthing`]
} [keyword else] {
	[keyword switch] [variable tag] {
	[keyword default]: [variable&callee s3]()
	[keyword case] [number 0], [number 1], [number 2], [number 3]: [variable&callee s1]()
	[keyword case] [number 4]: [keyword go] [variable&callee s2]()
	}
}

[def r], [def w] [operator :=] [variable os].[property&callee Pipe]([variable fd])
[def _], [def y], [def _] [operator :=] [type T]{[number 1], [number 3.44e+2]}

[keyword select] {
[keyword case] [variable i1] [operator =] [operator <-][variable c1]:
	[builtin print]([string "received "], [variable i1], [string " from c1\n"])
[keyword case] [variable c2] [operator <-] [variable i2]:
	[builtin print]([string "sent "], [variable i2], [string " to c2\n"])
[keyword case] [def&local i3], [def&local ok] [operator :=] ([operator <-][variable c3]):  [comment // same as: i3, ok := <-c3]
	[keyword if] [variable-2 ok] {
		[builtin print]([string "received "], [variable-2 i3], [string " from c3\n"])
	} [keyword else] {
		[builtin print]([string "c3 is closed\n"])
	}
[keyword case] [variable a][[[variable&callee f]()]] [operator =] [operator <-][variable c4]:
	[comment // same as:]
	[comment // case t := <-c4]
[keyword default]:
	[builtin print]([string "no communication\n"])
}

[comment /* this is a comment]
[comment   more comment]
[comment  */]

[def x] [operator :=] [type Foo]{
	[number 1], [variable&callee bar]([variable&callee foo]([variable a],
		[variable b])),
	[number 3], [number 4],
}

[keyword func] [def sieve]() {
	{
		[def&local m] [operator :=] [number 1];
	}
}

[keyword func] [def main]() {
	[keyword if] [def&local z] [operator :=] [number 1]; [variable-2 z] [operator %] [number 2] [operator ==] [number 1] {
		[def&local i] [operator :=] [number 2];
	} [keyword else] {
		[variable-2 z] [operator =] [number 2];
	}
	[variable fmt].[callee&property Println]([variable z])
}

[keyword func] ([def&local r] [type X]) [def y]() {
	[callee&variable bar]([variable-2 r], [builtin make]([type Bar]));
}

[keyword type] [def Message] [keyword struct] {
	[def&property Foo] [meta foo.][type string]
}

[keyword func] [def foo]() {
	[keyword var] [def&local wg] [meta sync.][type WaitGroup]
	[variable-2 wg].[property&callee Add]([number 1])
	[variable&callee f]([keyword func]() { [variable-2 wg].[property&callee Done]() })
	[variable-2 wg].[property&callee Wait]()
}

[keyword var] [def imported] [operator =] [keyword map][[ [type string] ]][type bool]{}

[keyword for] {
	[def&local x] [operator :=] [operator <-][variable ch]
}

[keyword type] [def A] [keyword interface] {
	[property Foo]([def&local f] [meta foo.][type Bar]) [keyword map][[ [type string] ]][type string]
	[property a]([def&local f] [meta foo.][type Bar])
	[type u]
}

[keyword type] [def Span] [keyword struct] {
	[type Range]
}

[callee&variable something]([keyword func]() { [keyword return] [number 22] })

[def abcd] [operator :=] [[]][meta bar.][type Baz]{
	{},
}
[keyword var] [def hashes] [operator =] [number 123]

[def abs] [operator :=] [[]][keyword struct] {
	[def&property a] [type string]
	[def&property b] [type int]
}{
	[number 22]
}

File: test/java/grammar.java
[keyword package] foo.bar.baz;

[keyword import] java.something.*;
[keyword import] java.blah.imported;

[comment /** multi]
[comment  @<i>N</i>]
[comment  * line]
[comment  * comment ][comment&tag @tag][comment */]

[meta @Foo](x = [string "something"])
[meta @Bar]
[meta @baz.argh]([variable Something].[property OK])
[keyword public] [keyword class] [type&def FontChoice] [keyword extends] [type Choice]<[type Param][[]]> {
  [comment // line comment]
  [keyword private] [keyword final] [keyword volatile] [meta java.blah.][qualified&type Font] [def&property fonts][[]];
  [keyword public] [def&property FontChoice]([type String] [def&local in], [type Globals] [def&local g]) {
    [keyword super]([variable-2 g]);

    [variable totalwidth] [operator =] [number 30];
    [type int] [def&local i], [def&local last] [operator =] [number 0b0_1], [def&local j] [operator =] [number 0.33e+3f];

    [keyword for] ([variable-2 i] [operator =] [number 0]; [variable-2 i] [operator <] [variable-2 in].[property&callee length](); [variable-2 i][operator ++])
      [keyword if] ([variable-2 in].[property&callee charAt]([variable-2 i]) [operator ==] [string ' '])
        [variable numofstyles][operator ++];
      [keyword else]
        [keyword throw] [keyword new]<[type Ugh], ?> [type Error]([string "double \" quoted"]);
    [keyword if] ([variable numofstyles] [operator ==] [variable foo][[ [number 0] ]])
      [keyword return];

    [variable names] [operator =] [keyword new] [type String][[ [variable numofstyles] ]];
    [keyword return] ([type Bar]<[type K], [type V]>) [variable&callee baz]();
  }

  [keyword protected] [type void] [def&property setX]([type int] [def&local x]) {
    [variable g] [operator ||] [variable-2 x];
    [keyword super].[property&callee setX]([variable-2 x]);
    [type FontMetrics]<[type Param]> [def&local fm] [operator =] [variable parent].[property&callee getFontMetrics]([variable fonts][[ [variable i] ]]);

    label: [keyword try] {
      [keyword continue] label;
    } [keyword catch] ([type RuntimeException] [def&local e]) {
      [variable system].[property&callee println]([variable-2 e].[property message]);
    } [keyword catch] ([keyword final] [type Exception] | [type InterruptedException] [def&local e]) {
    } [keyword finally] {
      [keyword return] [number 00l];
    }

    [variable foo].[property&callee addListeners](
        ([def&local x], [def&local y]) [operator ->] [atom true],
        [def&local z] [operator ->] [atom false] [operator ?] [variable-2 z] [operator :] [atom null]
        );
  }

  [def&property join]([type Bar]... [def&local bars]) {
  }

  [keyword enum] [type&def TrafficLight] {
    [def&property RED]([number 30]), [def&property AMBER]([number 10]),;

    [keyword private] [keyword final] [type int] [def&property seconds];

    [def&property TrafficLight]([type int] [def&local seconds]) {
      [keyword this].[property seconds] [operator =] [variable-2 seconds];
    }

    [type int] [def&property getSeconds]() {
      [keyword return] [variable seconds];
    }
  }
}

[keyword class] [type&def Foo] {
  <[type&def X]> [type void] [def&property foo]([type&local X] [def&local x], [type Y] [def&local y]) {
  }
}

[meta @BindingAnnotation]
[keyword @interface] [def SomeInterface] { }
[keyword public] [type void] [def run]() {
  [keyword return] [variable Foo].[keyword class];
}
